File: quest-sys/src/environment.cpp
//
// Created by Erich Essmann on 12/03/2025.
//
#include "bindings.h"

namespace quest_sys {
    void initQuESTEnv() {
        ::initQuESTEnv();
    }
    void initCustomQuESTEnv(int useDistrib, int useGpuAccel, int useMultithread) {
        ::initCustomQuESTEnv(useDistrib, useGpuAccel, useMultithread);
    }
    void finalizeQuESTEnv() {
        ::finalizeQuESTEnv();
    }
    void syncQuESTEnv() {
        ::syncQuESTEnv();
    }
    void reportQuESTEnv() {
        ::reportQuESTEnv();
    }
    bool isQuESTEnvInit() {
        return ::isQuESTEnvInit();
    }
    std::unique_ptr<QuESTEnv> getQuESTEnv() {
        return std::make_unique<QuESTEnv>(::getQuESTEnv());
    }

}

File: quest-sys/src/environment.rs
#[cxx::bridge]
pub mod ffi {
    // Types
    unsafe extern "C++" {
        // Opaque QuEST types
        type QuESTEnv;

        // Common type
        type Quest_Complex = crate::types::ffi::Quest_Complex;
    }

    // Environment management
    #[namespace = "quest_sys"]
    unsafe extern "C++" {
        include!("environment.hpp");
        fn initQuESTEnv();
        fn initCustomQuESTEnv(useDistrib: bool, useGpuAccel: bool, useMultithread: bool);
        fn finalizeQuESTEnv();
        fn syncQuESTEnv();
        fn reportQuESTEnv();
        fn isQuESTEnvInit() -> bool;
        fn getQuESTEnv() -> UniquePtr<QuESTEnv>;
    }
}
